### PR TITLE
Add `expect_and_forces` method to variational states

### PR DESCRIPTION
**Pull request by femtobit**
Gradients of observables are estimated in VMC via the covariance $F_j = \mathrm{cov}(O_j, E_\mathrm{loc})$ where $O_j = \partial\ln\psi(s)/\partial\theta_j$ is the derivative of the log-probability. This is called the _force vector_ here.

For real parameters, the observable gradient is then $\partial\langle \hat O\rangle / \partial\theta_j = 2\mathrm{Re}[F_j]$. For a complex holomorphic parametrization $\partial\langle \hat O\rangle / \partial\theta_j^* = F_j.$

Note that $F$ is generally complex even for real-parameter models. Since `expect_and_grad` returns the gradient and thus only the real part of $F$, information is lost in this step, which is needed in particular in time evolution (where the RHS of the equation of motion given by McLachlan's variational principle involves $\mathrm{Im}[F_j]$).

`expect_and_forces` adds a way to get the original forces for use in `TDVP`. I've added some tests and docs, the specific implementation in this PR has been written by @PhilipVinc.

### CODE REVIEW

File: netket/experimental/driver/tdvp.py [hunk 1]
```diff
@@ -532,6 +539,48 @@ def odefun(state: MCState, driver: TDVP, t, w, *, stage=0):  # noqa: F811
     return driver._dw
 
 
+@partial(jax.jit, static_argnums=(3, 4))
```
Comment (gcarleo) [inline netket/experimental/driver/tdvp.py:542]: what does this function do?

Comment (femtobit) [reply netket/experimental/driver/tdvp.py:542]: Essentially computing the RHS of the ODE from the forces. Its `Re[-i F]` for the real-time real-parameter case, `-i F` for real-time complex parameters, etc.

Maybe we can generalize it in the future to also support `Im[-i F]` (TDVP instead of McLachlan's VP) and similar stuff.

File: netket/vqs/mc/mc_state/expect_forces.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    # Code is a bit more complex than a standard one because we support
+    # mutable state (if it's there)
+    is_mutable = mutable is not False
+    _, vjp_fun, *new_model_state = nkjax.vjp(
+        lambda w: model_apply_fun({"params": w, **model_state}, σ, mutable=mutable),
+        parameters,
+        conjugate=True,
+        has_aux=is_mutable,
+    )
+    Ō_grad = vjp_fun(jnp.conjugate(O_loc) / n_samples)[0]
```
Comment (attila-i-szabo) [inline netket/vqs/mc/mc_state/expect_forces.py:104]: `vjp` returns the VJP cast on the dtype of the input, so doesn't this line yield only the real part of the force for real gradients?

**Comment by attila-i-szabo** [comment]
As a more general point, is there any use case where we need the real AND imaginary parts of the force vector at the same time? Otherwise, it'd be more efficient to only backprop `1` or `-1j` as needed

File: netket/experimental/driver/tdvp.py [hunk 3]
```diff
@@ -532,6 +539,48 @@ def odefun(state: MCState, driver: TDVP, t, w, *, stage=0):  # noqa: F811
...
+    #        raise ValueError(
+    #            "only real-time Lindblad evolution is supported for " "mixed states"
+    #        )
+    # else:
+    #    if propagation_type == "real":
+    #        loss_grad_factor = -1.0j
+    #    elif propagation_type == "imag":
+    #        loss_grad_factor = -1.0
+    #    else:
+    #        raise ValueError("propagation_type must be one of 'real', 'imag'")
```
Comment (PhilipVinc) [inline netket/experimental/driver/tdvp.py:567]:
```suggestion
```
can you add a docstring here?

File: netket/vqs/mc/mc_state/expect_forces.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+    # Ō_grad = jax.tree_map(
+    #    lambda x, target: (x if jnp.iscomplexobj(target) else 2 * x.real).astype(
+    #        target.dtype
+    #    ),
+    #    Ō_grad,
+    #    parameters,
+    # )
+
```
Comment (PhilipVinc) [inline netket/vqs/mc/mc_state/expect_forces.py:113]:
```suggestion
```

File: netket/vqs/mc/mc_state/expect_forces_chunked.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    )[0]
+
+    # Ō_grad = jax.tree_map(
+    #    lambda x, target: (x if jnp.iscomplexobj(target) else 2 * x.real).astype(
+    #        target.dtype
+    #    ),
+    #    Ō_grad,
+    #    parameters,
+    # )
+
```
Comment (PhilipVinc) [inline netket/vqs/mc/mc_state/expect_forces_chunked.py:155]:
```suggestion
```

**Review by PhilipVinc — APPROVED** [review]

**Comment by femtobit** [comment]
@attila-i-szabo As we discussed offline, most of the use cases involving the forces likely could be covered by moving away from `nkjax.vjp`. Computing the full force vector and then using only real or imaginary part does perform extra calculations (although without a benchmark it is hard to tell how much overhead that is) and we are open to adding a flag to allow `expect_and_forces` to only return the real or imaginary part forces then.

However, this can be done gradually (and based on our tests, nkjax does work correctly were it is used today), so we should be able to merge this PR right now and then build on it.

**Comment by gcarleo** [comment]
I am also worried about duplicating a lot of code by merging this, isn't there a way to make this the default logic and specialize it when it comes to getting only the real part at least?

File: netket/vqs/exact/expect.py [hunk 6]
```diff
@@ -63,20 +64,22 @@ def expect(vstate: ExactState, Ô: DiscreteOperator) -> Stats:  # noqa: F811
...
-    use_covariance: TrueT,
     *,
     mutable: Any,
 ) -> Tuple[Stats, PyTree]:
+    if isinstance(Ô, Squared):
+        raise NotImplementedError("expect_and_forces not yet implemented for `Squared`")
+
     _check_hilbert(vstate, Ô)
 
     O = sparsify(Ô)
```
Comment (gcarleo) [inline netket/vqs/exact/expect.py:78]: what does this do? can we add comments?

File: netket/vqs/mc/mc_state/expect_forces.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    )
+
+    if mutable is not False:
+        vstate.model_state = new_model_state
+
+    return Ō, Ō_grad
+
+
+@partial(jax.jit, static_argnums=(0, 1, 2))
+def forces_expect_hermitian(
```
Comment (gcarleo) [inline netket/vqs/mc/mc_state/expect_forces.py:67]: why? can we add comments?

File: netket/vqs/mc/mc_state/expect_grad.py [hunk 8]
```diff
@@ -39,41 +41,38 @@
 from .state import MCState
 
 
-@dispatch
-def expect_and_grad(  # noqa: F811
+@expect_and_grad.dispatch
+def expect_and_grad_covariance(
```
Comment (gcarleo) [inline netket/vqs/mc/mc_state/expect_grad.py:45]: what is this function doing? what case does it apply to? can we add comments?

**Comment by gcarleo** [comment]
This is a broader topic for discussion I guess, but more generally, I am a bit worried about the combinatorial proliferation of functions that do almost the same things, maybe at some point we should try to see if we can find a sustainable solution to this problem... I know dispatch is useful, but it is really hard to understand the inner parts of the code and why all the cases exist, say why there are :  

'forces_expect_hermitian_chunked', 'expect_and_grad_covariance', 'expect_and_grad_covariance_chunked' etc, or why 'Squared' is treated specially in the code has a few special branches (why not 'Cube' ?) etc

**Comment by femtobit** [comment]
> This is a broader topic for discussion I guess, but more generally, I am a bit worried about the combinatorial proliferation of functions

I agree with the general sentiment. The expectation value logic at the moment is distributed over lots of dispatching functions which do make it hard to follow (and modify) sometimes.

If I recall my discussions with @PhilipVinc correctly, there is redundancy between the chunked and non-chunked code which could be removed by merging them. That we might want to do (in a separate PR).

However, note that this PR mostly changes the old internal expect_and_grad functions (in all their variants) to return the forces (which are defined as a covariance) instead of gradients. So this introduces no additional complexity in that underlying code.

The API not contains two very similar functions expect_and_grad and expect_and_forces, but the grad one is directly implemented in terms of the forces, so while this adds some boilerplate, it introduces little additional logic.

In any case, I do agree this PR could be a good place to add some additional comments on internal functions.

**Comment by PhilipVinc** [comment]
As I said many times, I'm all ears for a new design that simplifies this logic AND passes all our existing tests. I do agree the current design is a bit convoluted.

However, If you try to implement something, you'd find out it's not easy, and you need some form of multiple dispatch or similar mechanism exactly to avoid duplicating code.

I thought long and hard about how and found no solution. Probably it can be done differently, and I know people don't like dispatch, but I think that
 - there's a reason why PyTorch is now moving all their internals to a multiple dispatch based system. It's the best solution to this kind of problem
 - I believe most of the pain and criticism here is that you are not familiar with multiple dispatch. This could be solved by writing documentation about it and tutorials and providing better errors.
 
I think the solution is probably in the middle between more documentation and less (but still some) dispatch. But someone needs to take a few weeks off from any other projects and work on this full time.

**Comment by PhilipVinc** [comment]
Aside: the chunking logic is duplicated because of a bug in jax or some imperfect thing in our implementation, which lead to worse performance when using the chunked code-path with no chunking.

We either devote more time of someone who knows jax well to optimise or implement workarounds in our code, or we keep the two code paths.

As nobody wanted to spend more time on hardcode reverse engineering the jax compiler, we decided to duplicate all implementations.

**Comment by gcarleo** [comment]
> I believe most of the pain and criticism here is that you are not familiar with multiple dispatch. This could be solved by writing documentation about it and tutorials and providing better errors.

I speak for myself of course, but I believe I understand what multiple dispatch does --essentially it's a glorified `if` on types, not really rocket science-- or is there more to it? The reason why I don't like it is because, contrary to the abstract class/function model, there is no **clear** way of knowing what are the APIs a function to be dispatched should obey to. One has to look deep into the code, often in "random" places, for similarly named functions and then somehow design a function that hopefully is compatible with the "mother" dispatch. In the abstract model setting (that we use in all other parts of netket, essentially) instead we know that by looking at the abstract/base class we have a clear idea of what the derived methods should do.  

Of course I understand that overall the problem is very hard to solve, and I find that the solution you put together is by many standards excellent, I am not saying otherwise. I am just hoping however we can maybe brainstorm all together at some point, writing down what are the requirements and the desired interface, and maybe we can come up with a solution more of us are comfortable with.

File: netket/vqs/exact/expect.py (hunk 6 above)
Comment (femtobit) [reply netket/vqs/exact/expect.py:78]: `sparsify` calls `O.to_sparse()` in order to evaluate the expectations via full summation. The function caches the last five evaluations in order to reduce the number of calls to `to_sparse`. I think the comment and docstring on `sparsify` are the best place to mention that, at least I'd keeping that information there once rather than repeating it at the places where `sparsify` is called:
https://github.com/netket/netket/blob/cdb85c77922efb7625f4281658c96f7326fb36aa/netket/vqs/exact/expect.py#L39-L46

**Comment by femtobit** [comment]
@PhilipVinc @gcarleo I've added some comments to hopefully make it clearer what each version of the function is for.

Is there anything else blocking this PR or can we merge it now?

**Comment by PhilipVinc** [comment]
I'm on holiday. I will officially get back on the 16 august. I will carefully look at it then

**Review by PhilipVinc — APPROVED** [review]

**Comment by PhilipVinc** [comment]
There are still some kinks to iron out, but let's release this
